test/e2e/specs/site-editor/font-library.spec.js

```diff
@@ -68,22 +68,183 @@ test.describe( 'Font Library', () => {
 		test( 'should show font variant panel when clicking on a font family', async ( {
 			page,
 		} ) => {
-			await page.getByRole( 'button', { name: /styles/i } ).click();
+			await page.getByRole( 'button', { name: 'Styles' } ).click();
 			await page
-				.getByRole( 'button', { name: /typography styles/i } )
+				.getByRole( 'button', { name: 'Typography Styles' } )
 				.click();
 			await page
 				.getByRole( 'button', {
 					name: /manage fonts/i,
 				} )
 				.click();
-			await page.getByRole( 'button', { name: /system font/i } ).click();
+			await page.getByRole( 'button', { name: 'System Font' } ).click();
+			await expect(
+				page.getByRole( 'heading', { name: 'System Font' } )
+			).toBeVisible();
 			await expect(
-				page.getByRole( 'heading', { name: /system font/i } )
+				page.getByRole( 'checkbox', { name: 'System Font Normal' } )
 			).toBeVisible();
+		} );
+	} );
+
+	test.describe( 'When user has permissions to edit font families', () => {
+		// Timeouts are increased in the following tests to account for the time it takes to load the font library tabs.
+
+		test.beforeAll( async ( { requestUtils } ) => {
+			await requestUtils.activateTheme( 'twentytwentythree' );
+		} );
+
+		test.beforeEach( async ( { admin, editor, page } ) => {
+			await admin.visitSiteEditor();
+			await editor.canvas.locator( 'body' ).click();
+			await page.getByRole( 'button', { name: 'Styles' } ).click();
+			await page
+				.getByRole( 'button', { name: 'Typography Styles' } )
+				.click();
+			await page
+				.getByRole( 'button', {
+					name: 'Manage Fonts',
+				} )
+				.click();
+		} );
+
+		test( 'should display the "Upload" tab and upload a local font', async ( {
+			page,
+		} ) => {
+			await expect(
+				page.getByRole( 'tab', { name: 'Upload' } )
+			).toBeVisible( { timeout: 40000 } );
```

Comment:
What condition is applying here in the UI to cause you to need such a long timeout?

Reply:
It seems to take quite a while to fetch the font family post data, it looks like it's the `useEntityRecords` call [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/global-styles/font-library-modal/context.js#L68) that takes a while to resolve. There doesn't seem to be any errors, just that it takes longer than the default timeout set in Playwight to resolve.

I'm currently looking at adding logic that would wait for this to resolve before running any of these tests. This should remove the need for specific timeout flags, but I'm assuming realistically it's still going to be a slower-than-average suite of tests, which I'm not a fan of adding.

Reply:
In which case I would be in favour of adding perhaps 1 or 2 high level "smoke tests". We might not catch every single bug but it might be enough to guard regressions around critical functionality (e.g. whether or not we display the `Upload` tab to users for example).

Reply:
What's the minimum set of tests we could introduce? Or are we already at that point?

Reply:
> In which case I would be in favour of adding perhaps 1 or 2 high level "smoke tests".

I've been thinking this too, I'll update the PR so the tests don't include uploading a font. We can always come back to that later if necessary. Also, I learned how to upload files in Playwright, so that's something 😄

> What's the minimum set of tests we could introduce? Or are we already at that point?
I like the idea of testing whether the Upload tab should appear or not based on permissions, so perhaps these are enough for now. It doesn't just test the permissions but also whether the font face posts have been initialised/fetched correctly, as the tab will only show once the posts have been fetched.

Reply:
> I learned how to upload files in Playwright, so that's something 😄

🎉 

> I like the idea of testing whether the Upload tab should appear or not based on permissions, so perhaps these are enough for now

This sounds like enough coverage and provides a good balance of test "weight" vs confidence.

Reply:
Thanks for your help here. I've reduced the tests to cover just the Upload tab display logic: https://github.com/WordPress/gutenberg/pull/59316/commits/c813923d7d5f18175271f7a1e4c76c15cec15e9d.

Reply:
> In which case I would be in favour of adding perhaps 1 or 2 high level "smoke tests".

Brainstorming a couple ideas

- An installed and activated font is shown correctly in the list of Font Library fonts
- The Google Fonts collection has a tab shown in the Font Library modal

> I like the idea of testing whether the Upload tab should appear or not based on permissions

I see how it makes sense to use e2e testing on this (check that changing a server side setting is reflected correctly in the client). But using custom permissions seems like a rare use case (I'd guess a small fraction of a percent of sites), so still not sure it's worth spending an e2e test on.

Reply:
I've slightly lost track of the discussion in other tickets now (it's evolving quickly) but it sounded to me that it was very important that the UI did not allow uploading if the server settings or permissions did not allow for it.

If it's of significant enough important for people to raise flags right before Beta 3 of a major WP release, then it feels like we should provide some level of coverage to avoid future regresions.

That said, maybe it's not custom permissions we care about but rather the `DISALLOW_FILE_MODS` constant?

If you all feel it's totally unneeded then I'll defer to you as the contributor's who've been actively involved in the feature.

Reply:
> An installed and activated font is shown correctly in the list of Font Library fonts
The Google Fonts collection has a tab shown in the Font Library modal

These were covered previously but I've just removed them to simplify these tests, but happy to add them back! The main issue I'm seeing is that resolving the font family posts takes a few extra seconds, so this means we'd be adding some potentially slow e2e tests. I'm wondering if there's something we can improve here in the order in which the font posts are requested, i.e. perhaps the request could start at the same time as the modal opening, rather than after the modal has opened.

> But using custom permissions seems like a rare use case (I'd guess a small fraction of a percent of sites), so still not sure it's worth spending an e2e test on.

Maybe we should test whether the Font Library UI is disabled via the editor settings rather than checking permissions. Does that sound worthy of a test?

Reply:
> Maybe we should test whether the Font Library UI is disabled via the editor settings rather than checking permissions. Does that sound worthy of a test?

Yes, I think it would be.

Reply:
> The main issue I'm seeing is that resolving the font family posts takes a few extra seconds, so this means we'd be adding some potentially slow e2e tests.

Not sure if this is a good idea, but it should be possible to preload the font-families API response using a filter, in the same way that essential routes that are used when the editor first loads are preloaded, by outputting the response JSON into the html markup on the page when it loads.

But I don't think this would make sense normally, where the font library may or may not be opened. It would only be for the test environment--hence not being sure if it's a good idea to "fork" the test environment like this. But it may be worth it--the only change is where the response data comes from, the actual API response should be exactly the same either way.

Reply:
Can we just eliminate the default font collection from the tests?  Unregister that one and register another, much smaller one that won't require a json file to resolve at all.  We'll probably wind up with a few different font collection scenarios to manufacture and test anyway.

Reply:
> Not sure if this is a good idea, but it should be possible to preload the font-families API response using a filter

There are a few instances in the e2e tests where we wait for either certain blocks to load in or specific endpoints to resolve before moving on, so I think we can use a similar method here.

> Can we just eliminate the default font collection from the tests?

The default font collection isn't loaded until the Google Fonts acceptance notice has been clicked, which hasn't been tested so far, so the whole collection hasn't been resolved as part of these tests, only the tab.

It's the initial post request that seems to take a while, but I'm not sure what part of it is taking a while or if it's a sign that performance needs to be improved somewhere. I usually find it easy to replicate in a Playground instance too - the initial request takes a noticeable 5-8 seconds. I'm not sure if addressing this is an urgent priority, and initially, perhaps we could look at improving the UX around it. Currently, there is a spinner that appears during the post request and then disappears once resolved. If there are no custom fonts, it's not clear what the spinner is indicating.

Also apologies as I think I'm deviating a bit here 😅 I plan to finish up the minimal tests in this PR and maybe we should open a separate issue to discuss performance improvements.